src/modules/users/users.controller.ts

```diff
@@ -31,19 +31,90 @@ export class UsersController {
   @ApiOperation({ summary: 'Retrieve all admins' })
   @ApiResponse({ status: 200, description: 'List of all admins', type: [User] })
   @HttpCode(HttpStatus.OK)
-  findAll(
+  findAllAdmins(
     @Body() paginationUserDto: PaginationUserDto,
   ): Promise<{ data: User[]; total: number }> {
-    return this.usersService.findAndPaginateAllAdmins(paginationUserDto);
+    const allowedRole: UserRole = UserRole.ADMIN;
+
+    return this.usersService.findAndPaginateAllUsersWithRole(
+      paginationUserDto,
+      allowedRole,
+    );
+  }
+
+  @Post('/dispatchers')
+  @Roles(UserRole.ADMIN)
+  @ApiOperation({ summary: 'Retrieve all dispatchers' })
+  @ApiResponse({
+    status: 200,
+    description: 'List of all dispatchers',
+    type: [User],
+  })
+  @HttpCode(HttpStatus.OK)
+  findAllDispatchers(
+    @Body() paginationUserDto: PaginationUserDto,
+  ): Promise<{ data: User[]; total: number }> {
+    const allowedRole: UserRole = UserRole.DISPATCHER;
+
+    return this.usersService.findAndPaginateAllUsersWithRole(
+      paginationUserDto,
+      allowedRole,
+    );
+  }
+
+  @Post('/drivers')
+  @Roles(UserRole.ADMIN, UserRole.DISPATCHER)
+  @ApiOperation({ summary: 'Retrieve all drivers' })
+  @ApiResponse({
+    status: 200,
+    description: 'List of all drivers',
+    type: [User],
+  })
+  @HttpCode(HttpStatus.OK)
+  findAllDrivers(
+    @Body() paginationUserDto: PaginationUserDto,
+  ): Promise<{ data: User[]; total: number }> {
+    const allowedRole: UserRole = UserRole.DRIVER;
+
+    return this.usersService.findAndPaginateAllUsersWithRole(
+      paginationUserDto,
+      allowedRole,
+    );
   }
 
   @Get('/admin/:id')
   @Roles(UserRole.SUPER_ADMIN, UserRole.ADMIN)
   @ApiOperation({ summary: 'Retrieve a single admin by ID' })
   @ApiResponse({ status: 200, description: 'Admin found', type: User })
   @HttpCode(HttpStatus.OK)
-  findOne(@Param('id', ParseIntPipe) id: number): Promise<User | null> {
-    return this.usersService.findOneAdmin(id);
+  findOneAdmin(@Param('id', ParseIntPipe) id: number): Promise<User | null> {
+    const allowedRole: UserRole = UserRole.ADMIN;
+
+    return this.usersService.findOneUser(id, allowedRole);
+  }
+
+  @Get('/dispatcher/:id')
+  @Roles(UserRole.ADMIN)
+  @ApiOperation({ summary: 'Retrieve a single dispatcher by ID' })
+  @ApiResponse({ status: 200, description: 'Dispatcher found', type: User })
+  @HttpCode(HttpStatus.OK)
+  findOneDispatcher(
+    @Param('id', ParseIntPipe) id: number,
+  ): Promise<User | null> {
+    const allowedRole: UserRole = UserRole.DISPATCHER;
```

Comment:
why this extra varuable needed?

Reply:
For reusability of functions in the user service.